osu.Framework/Input/States/TouchState.cs

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+
+namespace osu.Framework.Input.States
+{
+    public class TouchState
+    {
+        /// <summary>
+        /// Represents a list of currently active pointers.
+        /// </summary>
+        public readonly ButtonStates<PositionalPointer> Pointers = new ButtonStates<PositionalPointer>();
+
+        /// <summary>
+        /// Attempts to retrieve the recent pointer of all active pointers.
+        /// </summary>
+        /// <returns>Whether the pointer is successfully retrieved.</returns>
+        public bool TryGetPrimaryPointer(out PositionalPointer primary)
+        {
+            try
+            {
+                primary = Pointers.OrderBy(p => p.Source).Last();
```

Comment:
Will there be same sources?
```suggestion
                primary = Pointers.OrderByDescending(p => p.Source).First();
```

Reply:
There shouldn't be. (check `ButtonsState.SetPressed()`)

is there a difference between last ordered by descending and first ordered by ascending?

Reply:
No, but `First` and `Last` on ordered enumerable will be different if the implementation is not smart enough.
After looking into linq source, the current master branch has a special path for it. But I still prefer `First()`.